R/thematic.R

```diff
@@ -132,31 +111,87 @@ thematic_off <- function() {
   ggplot_build_restore()
   lattice_print_restore()
 
-  if (!is.null(.globals$theme)) rm("theme", envir = .globals)
+  theme <- .globals$theme
+  if (!is.null(theme)) rm("theme", envir = .globals)
+
+  invisible(theme)
+}
+
+#' @rdname thematic
+#' @export
+thematic_theme <- function(bg = "auto", fg = "auto", accent = "auto",
+                           font = NA, sequential = sequential_gradient(),
+                           qualitative = okabe_ito(), inherit = FALSE) {
+
+  # This function is called at plot time (with bg/fg/accent)
+  if (is.function(sequential)) {
+    sequential <- structure("", sequential_func = sequential)
+  }
+
+  new <- structure(
+    list(
+      bg = tag_auto(bg),
+      fg = tag_auto(fg),
+      accent = tag_auto(accent),
+      qualitative = qualitative,
+      sequential = sequential,
+      font = as_font_spec(font)
+    ),
+    class = "thematic_theme"
+  )
+
+  # Newly _specified_ theme elements override old elements
+  old <- thematic_get_theme()
+  if (isTRUE(inherit) && length(old)) {
+    fmls <- formals(thematic_theme)
+    if (identical(bg, fmls$bg)) new$bg <- NULL
+    if (identical(fg, fmls$fg)) new$fg <- NULL
+    if (identical(accent, fmls$accent)) new$accent <- NULL
+    if (identical(sequential, fmls$sequential)) new$sequential <- NULL
+    if (identical(qualitative, fmls$qualitative)) new$qualitative <- NULL
+    new <- utils::modifyList(old, new)
+    new <- structure(new, class = "thematic_theme")
+  }
 
-  invisible()
+  new
 }
 
+is_thematic_theme <- function(x) {
+  inherits(x, "thematic_theme")
+}
+
+#' @rdname thematic
+#' @param session see [shiny::onStop()].
+#' @export
+thematic_shiny <- function(..., session = shiny::getDefaultReactiveDomain()) {
+  old_theme <- thematic_on(...)
+  shiny::onStop(function() thematic_set_theme(old_theme), session = session)
+  invisible(old_theme)
+}
```

Comment:
Maybe there should also be a `thematic_rmd()` which cleans up after itself in a knitr exit hook